### PR TITLE
add support for external json files

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Solve issue #5, #6, Avaiga/taipy-studio#133

### CODE REVIEW

File: src/assets/find_element_file.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from importlib import util
+import inspect
+import os
+
+if util.find_spec("taipy") and util.find_spec("taipy.gui"):
+    from taipy.gui import Gui
+    element_file_path = f"{os.path.dirname(inspect.getfile(Gui))}{os.sep}webapp{os.sep}viselements.json"
```
Comment (FredLL-Avaiga) [inline src/assets/find_element_file.py:7]: Any specific reason why is the viselements.json file in webapp?

File: src/gui/command.ts [hunk 2]
```diff
@@ -62,7 +63,8 @@ export class GenerateGuiCommand {
 
     private static async handleElementPropertySelection(guiElement: GuiElement) {
         const quickPick = window.createQuickPick();
-        const propertyObject = defaultElementProperties[guiElement.elementName as keyof typeof defaultElementProperties];
+        const elementProperties = ElementProvider.getElementProperties();
+        const propertyObject = elementProperties[guiElement.elementName as keyof typeof elementProperties];
         if (Object.keys(propertyObject).length === 0) {
```
Comment (FredLL-Avaiga) [inline src/gui/command.ts:68]: can propertyObject be undefined ?

File: src/gui/command.ts [hunk 3]
```diff
@@ -97,3 +99,49 @@ export class GenerateGuiCommand {
...
+    private constructor(readonly context: ExtensionContext) {
+        context.subscriptions.push(
+            commands.registerCommand("taipy.gui.md.findElementFile", FindElementsFileCommand.commandEntry)
+        );
+    }
+
+    private static async commandEntry() {
+        const result = await window.showInputBox({
+            placeHolder: l10n.t("Enter Python executable path. `python` value will be used if the field is empty"),
+            prompt: l10n.t("This will be used to determine your elements file from taipy-gui package"),
```
Comment (FredLL-Avaiga) [inline src/gui/command.ts]: elements file from => elements file path from

File: src/gui/command.ts [hunk 4]
```diff
@@ -97,3 +99,49 @@ export class GenerateGuiCommand {
...
+    }
+
+    private constructor(readonly context: ExtensionContext) {
+        context.subscriptions.push(
+            commands.registerCommand("taipy.gui.md.findElementFile", FindElementsFileCommand.commandEntry)
+        );
+    }
+
+    private static async commandEntry() {
+        const result = await window.showInputBox({
```
Comment (FredLL-Avaiga) [inline src/gui/command.ts:115]: could we not try and check if python exists before asking for a path ?

File: src/gui/command.ts [hunk 5]
```diff
@@ -97,3 +99,49 @@ export class GenerateGuiCommand {
...
+                    let execResult = await execShell(
+                        `${result || "python"} ${join(__dirname, "assets", "find_element_file.py")}`
+                    );
+                    if (execResult.startsWith("Path: ")) {
+                        updateFilePath(execResult.substring(6));
+                        window.showInformationMessage(l10n.t("Element Path has been found and updated in workspace settings"));
+                    } else if (execResult) {
+                        window.showErrorMessage(execResult);
+                    } else {
+                        window.showErrorMessage(l10n.t("Can't find element file path with the provided environment"));
```
Comment (FredLL-Avaiga) [inline src/gui/command.ts]: element => elements ?

File: src/gui/command.ts [hunk 6]
```diff
@@ -97,3 +99,49 @@ export class GenerateGuiCommand {
...
+                title: "Finding Element File",
+            },
+            async (progress) => {
+                try {
+                    let execResult = await execShell(
+                        `${result || "python"} ${join(__dirname, "assets", "find_element_file.py")}`
+                    );
+                    if (execResult.startsWith("Path: ")) {
+                        updateFilePath(execResult.substring(6));
+                        window.showInformationMessage(l10n.t("Element Path has been found and updated in workspace settings"));
```
Comment (FredLL-Avaiga) [inline src/gui/command.ts]: Element Path => Elements file path ?

File: src/gui/constant.ts [hunk 7]
```diff
@@ -11,8 +11,10 @@
  * specific language governing permissions and limitations under the License.
  */
 
-import visualElements from "../assets/viselements.json";
-import { getBlockElementList, getControlElementList, getElementList, getElementProperties, getOnFunctionList, getOnFunctionSignature } from "./utils";
+import { join } from "path";
+import { getBlockElementList, getControlElementList, getElementFile, getElementList, getElementProperties, getOnFunctionList, getOnFunctionSignature } from "./utils";
+
+const visualElements = getElementFile(join(__dirname, "assets", "viselements.json")) || {};
```
Comment (FredLL-Avaiga) [inline src/gui/constant.ts:17]: I think __dirname is only available in a node.js context ie it is evaluated at transpile time
I suppose what you need to do is "require" the json file  so that Webpack can rewrite the path at transpile/compile time

File: src/gui/elementProvider.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        }
+        if (propertyName in this.elementCache[filePath]) {
+            return this.elementCache[filePath][propertyName];
+        }
+        const visualElements = this.elementCache[filePath]["visualElements"] || getElementFile(filePath);
+        if (!visualElements) {
+            return defaultValue;
+        }
+        if (!(propertyName in this.elementCache[filePath])) {
+            this.elementCache[filePath].visualElements = visualElements;
```
Comment (FredLL-Avaiga) [inline src/gui/elementProvider.ts]: Should we use the static access here ? this => ElementProvider

File: src/gui/utils.ts [hunk 9]
```diff
@@ -165,3 +167,50 @@ export const generateOnFunction = (signature: [string, string][], functionName:
...
+        if ("controls" in elements && "blocks" in elements && "undocumented" in elements) {
+            return elements;
+        }
+        // Reset if filepath is not valid
+        if (path === getElementFilePath()) {
+            updateFilePath("");
+        }
+        return undefined;
+    } catch (error) {
+        return undefined;
```
Comment (FredLL-Avaiga) [inline src/gui/utils.ts:215]: console.info ?

**Review by FredLL-Avaiga — APPROVED** [review]
__dirname needs a rewrite :-(

File: src/gui/constant.ts (hunk 7 above)
Comment (FredLL-Avaiga) [reply src/gui/constant.ts:17]: Sorry, I was wrong ..
as vscode is node.js based, dirname is defined 
But we'll have to check where it points to when the extension is deployed

File: l10n/bundle.l10n.json [hunk 10]
```diff
@@ -1,16 +1,25 @@
...
+  "Create function '{0}'": "Create function '{0}'",
+  "Replace with '{0}'": "Replace with '{0}'",
+  "Remove negated value": "Remove negated value",
   "Select an element type": "Select an element type",
   "Enter element value": "Enter element value",
   "Unmatched number of curly braces for expression": "Unmatched number of curly braces for expression",
   "Select properties for element '{0}'": "Select properties for element '{0}'",
   "Element added": "Element added",
+  "Enter Python executable path. `python` value will be used if the field is empty": "Enter Python executable path. `python` value will be used if the field is empty",
+  "This will be used to determine your elements file from taipy-gui package": "This will be used to determine your elements file from taipy-gui package",
```
Comment (FabienLelaquais) [inline l10n/bundle.l10n.json]: => "this path is used to locate the elements description file" 
... i think.

File: l10n/bundle.l10n.json [hunk 11]
```diff
@@ -1,16 +1,25 @@
...
+  "Element Path has been found and updated in workspace settings": "Element Path has been found and updated in workspace settings",
```
Comment (FabienLelaquais) [inline l10n/bundle.l10n.json]: Element Path, then element file path?

Give me a call

File: l10n/bundle.l10n.json [hunk 12]
```diff
@@ -1,16 +1,25 @@
...
+  "Can't find element file path with the provided environment": "Can't find element file path with the provided environment",
   "Missing closing syntax": "Missing closing syntax",
   "Missing opening tag": "Missing opening tag",
   "Missing matching opening tag identifier '{0}'": "Missing matching opening tag identifier '{0}'",
   "Missing matching closing tag identifier '{0}'": "Missing matching closing tag identifier '{0}'",
   "Unmatched opening tag identifier '{0}'": "Unmatched opening tag identifier '{0}'",
   "Unmatched closing tag identifier '{0}'": "Unmatched closing tag identifier '{0}'",
   "Missing closing tag with tag identifier '{0}'": "Missing closing tag with tag identifier '{0}'",
```
Comment (FabienLelaquais) [inline l10n/bundle.l10n.json]: Missing closing tag with tag identifier
->
Missing closing tag for opened tag

tag and identifier really mean the same thing
tag is what we use for element chunks.

File: package.json [hunk 13]
```diff
@@ -18,14 +18,17 @@
   "icon": "assets/taipy-logo.png",
   "l10n": "./dist/l10n",
   "categories": [
+    "Linters",
```
Comment (FabienLelaquais) [inline package.json:21]: 👍

File: package.nls.json [hunk 14]
```diff
@@ -1,3 +1,5 @@
 {
-    "taipy.gui.md.generate.title": "Generate Taipy GUI element"
+    "taipy.gui.md.generate.title": "Generate Taipy GUI element",
```
Comment (FabienLelaquais) [inline package.nls.json]: Generate Taipy GUI element
if it appears in the 'Taipy' category, I would rephrase that to "Generate visual element"

File: src/assets/find_element_file.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+    if os.path.exists(element_file_path):
+        print(f"Path: {element_file_path}")
+    else:
+        print("element file not found in taipy-gui package")
```
Comment (FabienLelaquais) [inline src/assets/find_element_file.py]: element -> Element description

**Review by FabienLelaquais — APPROVED** [review]
A few wording issues.
We need to be more consistent.

File: src/assets/find_element_file.py (hunk 1 above)
Comment (dinhlongviolin1) [reply src/assets/find_element_file.py:7]: because it is built into the webapp per the latest pr from taipy-gui

File: src/gui/command.ts (hunk 2 above)
Comment (dinhlongviolin1) [reply src/gui/command.ts:68]: It would not be empty since the user will need to select the element type before hand

File: src/gui/command.ts (hunk 6 above)
Comment (dinhlongviolin1) [reply src/gui/command.ts]: thanks!

File: src/gui/utils.ts (hunk 9 above)
Comment (dinhlongviolin1) [reply src/gui/utils.ts:215]: will do

File: src/gui/elementProvider.ts (hunk 8 above)
Comment (dinhlongviolin1) [reply src/gui/elementProvider.ts]: Thanks!

File: src/gui/command.ts (hunk 4 above)
Comment (dinhlongviolin1) [reply src/gui/command.ts:115]: we can definitely try that, but the point being that they can choose the python environment which have taipy-gui installed, or the one they are using for their workspace

File: package.nls.json (hunk 14 above)
Comment (dinhlongviolin1) [reply package.nls.json]: It does not, but we can try this `Taipy Studio Gui Helper: Generate Visual Element`
